### DISCUSSION TITLE
mosaic and use cutline to display only in specific area

### DISCUSSION THREAD

**Discussion by satas12825** [cogeotiff/rio-tiler#625] [Q&A] [answered]
I try to  mosaic two of cog tif files and then i want to apply cutline to it but it does not work

**Comment by vincentsarago** [comment]
@satas12825 please provide more context or code 🙏

**Reply by satas12825** [reply]
thank you for your response, I already find the solution but when I use cutline the border of the result was not smooth and how can I fix it? 
![image](https://github.com/cogeotiff/rio-tiler/assets/70628269/e5697837-6b3d-4def-a957-144dda8af438)

` 
# cutline
    with Reader(path3) as cog:
        cutline = create_cutline(cog.dataset,geojson,geometry_crs="epsg:4326")
      
    

    mosaic_assets = [path3,path4]

    img, _ = mosaic_reader(
        mosaic_assets,
        tiler,
        x,
        y,
        z,
        indexes=[4,3,2],
        vrt_options={'cutline': cutline},
        pixel_selection=defaults.HighestMethod()
       

    )

    img.rescale(
        in_range=((0, 10000),),
        out_range=((0, 255),),
    )


    # img.tile(z, x, y, vrt_options={'cutline': cutline})
    content = img.render(img_format="PNG", **img_profiles.get("png"))`

**Reply by vincentsarago** [reply]
pixels are squares... 

if you want only the pixel with center points within you shape, you can maybe change https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/io/rasterio.py#L577 to False (https://github.com/rasterio/rasterio/blob/main/rasterio/_features.pyx#L293-L296) 

maybe @yellowcap can explain it more 😄 (https://github.com/cogeotiff/rio-tiler/pull/598#discussion_r1179218090)

**Comment by yellowcap** [comment, marked as answer]
> maybe @yellowcap can explain it more smile (https://github.com/cogeotiff/rio-tiler/pull/598#discussion_r1179218090)

Yes the way the cutline is configured, all pixels that are touched by the cutline (i.e. the pixel boundary polygon). So pixels will tend to slightly "overflow" the cutline geometry lines. The reason for this is explained in the comment @vincentsarago shared. We want this cutline to give identical results, even if there are multiple resolution in ImageData. This is only guaranteed for `all_touched=True` mode, otherwise smaller pixels will be dropped and the masks will end up being different. 

One way to work around this for now would be to use a negative buffer of about 1 pixel size on your geometry before using it as cutline. This will make the rasterization more constraint.

Another one would be to increase the resolution of your target raster, then the cutline is going to be  following a finer grid, so that could help too.

**Comment by satas12825** [comment]
@yellowcap @vincentsarago 
I already got it, Thank you for taking the trouble to answer all my questions.
